### ISSUE TITLE
Duration of a test case includes setup/teardown

### ISSUE THREAD

**Issue by parsonsmatt** [hspec/hspec#633]
At work, we've got a function `withApp :: SpecWith (TestApp App) -> Spec` that grabs a fresh database connection from a pool of ~4 available connections. Tests are often started, but then spend some (significant) time waiting on the database pool to be available. This time appears to be reported in the `--print-slow-spec-items`, and it wreaked havoc on a `timeout` function we tried to implement on our test suite.

Ideally, the `Duration` for a test would only be on the actual test example itself, not all the stuff `before` or `after` it.

```haskell
spec :: Spec
spec = before (threadDelay 100000) $ it "should work" True
```

I dug around in the code for a bit last night but was unable to figure out precisely what is determining the duration, or if there's anything done to separate the various `IO` hooks and the eventual `Example a => a` stuff.

**Comment by sol** [comment]
Before discussing this in more detail, are you running your tests concurrently?  If yes, have you tried to limit concurrency**?

** Pass `--jobs=4` to your test driver (as an alternative, if your are on `hspec-2.8.4` or later, set the environment variable `HSPEC_JOBS=4`).

> it wreaked havoc on a `timeout` function we tried to implement on our test suite.

Can you elaborate? What did that timeout function do and how did that wreak havoc.

**Comment by parsonsmatt** [comment]
Yes, this seems to only occur in tests when we run them in parallel. Otherwise there's no contention on the database connection, and no time is spent waiting on the database connection to be acquired.

However, our `after` action involves deleting the database tables and getting back to a fresh start, and that *is* being reflected in the test duration - for our tests that touch the database, almost half of that is spent in wiping the tables clean in the `after` part, and that is being reflected in the test duration.

> Can you elaborate? What did that timeout function do and how did that wreak havoc.

We want to assert that a test does not take longer than 60s. Sometimes just acquiring a database connection would take that long in CI, and it'd kill the thread, which would have cascading weird behavior in the rest of the test suite. I don't think it's *super* necessary to worry about this particular thing, since the weird behavior is almost certainly internal to our use of `unsafePerformIO` to have the connection pools ready.

**Comment by sol** [comment]
> I dug around in the code for a bit last night but was unable to figure out precisely what is determining the duration, or if there's anything done to separate the various `IO` hooks and the eventual `Example a => a` stuff.

`befoer` / `after` is part of the spec item (aka example).  Hooks basically transform a spec item.  For QuickCheck properties, every invocation of the property is transformed.  This is why we pass around the transformation function and why the types get somewhat daunting.

If we try hard, we could probably get separate times for `beforeAll` / `afterAll` somehow (`afterAll` is already separate, but I was going to make it less separate in https://github.com/hspec/hspec/pull/604).  For per-item hooks, I'm not sure.


A very personal view: For me, personally, doing anything here is a "nice to have" rather than "critical".  Yes, it could be "nice" or "more correct".  But if I'm at the point that I care about this I'm already in big trouble.  Long setup / teardown (and tests that take a long time in general) significantly impact my productivity and that's what I avoid at all costs.

**Comment by sol** [comment]
> We want to assert that a test does not take longer than 60s. Sometimes just acquiring a database connection would take that long in CI

Did you try `--jobs=4` to address this.  If you limit the number of threads to the number of database connections I would assume that this should work as a "workaround".  Once we have verified that the workaround works, this is what I'm puzzled:

- What is the benefit of using more threads than database connections?  Does your test suite run significantly faster if you do that? (specifically, given `--jobs=n` and a thread pool of size `m`, will `n > m` give you any tangible performance benefit for your particular test suite?)
- What is the reason that you limit the number of database connection in the first place?  Does your test suite run significantly slower if you don't limit them? (I could assume that if you use PostgreSQL and have a lot of cores it could, at least historically this was an issue with Postgres, but I was under the impression that it is less of an issue with recent versions)

**Comment by sol** [comment]
> However, our `after` action involves deleting the database tables and getting back to a fresh start, and that _is_ being reflected in the test duration

Ok, so from what I understand, even if you limit concurrency, you still run into issues because (a) cleanup can take a long time and (b) the cleanup is included in the execution time of the spec item.

The ideal solution would be to exclude setup / teardown.  That would solve your issue and would possibly be easier to understand and benefit other users.

Let's put this aside for a moment, and focus on your issue for now.  How did you try to apply timeouts to your spec items? In a hook?  Could you apply the timeout inside your `withConnection` instead?  E.g. if you have something like:

```haskell
withConnection action = do
  withResource pool $ \ conn -> do
    action conn `finally` cleanup conn
```

change it to something like

```haskell
withConnection action = do
  withResource pool $ \ conn -> do
    withTimout n (action conn) `finally` cleanup conn
```

This has the limitation that the timeout is only applied to db specs (say, if you have other spec items that are slow then they won't timeout).

> for our tests that touch the database, almost half of that is spent in wiping the tables clean in the `after` part, and that is being reflected in the test duration

This issue aside + and at the risk of telling you something that you already know: I would try very hard to avoid this in the first place.  Wrap your tests in a transaction, `begin` / `rollback`.  That way you don't have to do any cleanup and the db does not spend cycles on committing changes that you are going to revert manually later on.  This will **require some changes to how you structure your application** and how you use transactions in your application code (PostgreSQL does not support nested transactions), but it will **significantly speed up your tests**.

For completeness, here is some pseudo code:

```haskell
withConnection action = do
  withResource pool $ \ conn -> do
    (begin conn >> action conn) `finally` rollback conn
```
(You will have to pass that very same `conn` into your app of course, and make sure that the app does not randomly acquire other connections.  One way to achieve this is to create a pseudo pool that always returns the same connection.  This works for most common cases, like controller code, but not if you do something weird, like randomly spawning worker threads or something...)

There are more facets to this + some people might conclude that it does not suite their *special situation* for some reason.  I think more often then not, this conclusion is premature.  There are really not many good reasons that justify having slow tests and by extension lower productivity and less fun at work every single day.  The only thing that would convince me is code that proofs the point.  So here is the deal, anybody, **show me your *special situation*, and I'll tell you how to make your tests fast**.

**Comment by hdgarrood** [comment]
I agree that, where possible, using transactions within tests is great. Unfortunately, this is unworkable for the situation we're in, which I think is pretty normal - we want to perform integration tests on our HTTP API, where a particular test case can involve calls to 2 or more endpoints in sequence. We want to be able to test things like "after hitting endpoint A with request body B, a request to endpoint C results in response D". I don't think there's any good way of having a HTTP API backend keep a single transaction open across two or more separate requests; even if we were able to make it work, to me, that would constitute an unacceptably significant change from how the app actually works in practice, such that I wouldn't have confidence in the tests.

**Comment by sol** [comment]
@hdgarrood I agree. It won’t work for user journey tests and some acceptance tests.

**Comment by sol** [comment]
> Ideally, the `Duration` for a test would only be on the actual test example itself, not all the stuff `before` or `after` it.

Revisiting this issue,  I understand that this is true for your situation. In other situations you are probably interested how much time is spent on some test case including before/after.

1. If we are going to do something here then I think we would not want to simply discard the before/after time but instead have a breakdown by before/test/after and corresponding options to control the output.
2. For QuickCheck properties, before/after is run for each invocation of the property. It’s not obvious to me how you would want to handle this.
3. I’m not aware of a simple way to implement this. I have some vague idea that would require to modify the `Example` type class, requiring custom code for each and every instance. @parsonsmatt if you’re still interested in this then I may do a proof of concept when I get to it. But all the legwork would need to be done by you, or some other volunteer.

**Comment by codygman** [comment]
We experienced this same issue in our company codebase.

Our workaround was to do `setupDatabases` before calling hspec. I guess the negative of that is hspec isn't the entry point to the tests, but things seem to work.

**Comment by codygman** [comment]
> Long setup / teardown (and tests that take a long time in general) significantly impact my productivity and that's what I avoid at all costs.

That's actually the reason why not having setup/teardown included in the slowspec time mattered to me.

My team is using slowspec as a "slow test regression" where any spec more than 200ms gets printed out. In practice this means developers adding tests know there is a problem if there is more than 5 lines of output on ghcid reload.

It also means the slowspec log is hidden once you've hit your goal of "all specs faster than 200ms".

I have a fork with that filtering ability I'm planning to PR soon that adds the ability to filter slow spec items by name and duration. In our codebase for instance we only want to see specs that take more than 300ms to begin with.

Then as we are able to spend more time improving the state of our 1000+ tests, I'd like to ratchet it down to 200ms or even 100ms.

> But if I'm at the point that I care about this I'm already in big trouble. 

This works for greenfield codebases but not legacy ones where tests can take minutes or hours and the slowspec by default prints out 2500 tests. It's much more manageable and politically viable to say "let's fix the top 2% of tests and make builds fail if new tests taking longer than those are added in the future".

**Comment by sol** [comment]
@parsonsmatt @hdgarrood @parsonsmatt I'm playing around with some code that looks promising.  I'll let you guys know once there is something that you can try out.

**Comment by sol** [comment]
#765 implements `--ignore-hook-times=all|global|none`. Please give it a try and provide feedback.

Note that you can set this option by different means (env var, `.hspec`, `modifyConfig` in `test/SpecHook.hs`...).

Questions I want to have answered:
1. Is this what you folks were looking for? Does this address your use cases?
2. Do we need both `global` and `all`? What are you primarily interested in using?
3. It will be possible to implement `--timeout` on top of this. @parsonsmatt Is this what you primarily care about, or do you care about `--print-slow` too?

**Comment by sellout** [comment]
I’ve run into what I think is a similar case – `describe "meh" $ around_ inTempDirectory $ do …` was breaking our hspec-golden tests, because `evaluateExample` for `Example (IO (Golden str))` expects to read/write the golden file.

My expectation would be that something like `around_` would apply to the action passed _to_ `evaluateExample`, not the context of the `Example`.

**Comment by sol** [comment]
> I’ve run into what I think is a similar case

Without any further evidence, I am not convinced that this is indeed the case.

@sellout please open a separate issue with exact steps on how to reproduce.

**Comment by sellout** [comment]
> > I’ve run into what I think is a similar case
> 
> Without any further evidence, I am not convinced that this is indeed the case.

You’re probably right, I was just seeing how they both involve scoping of hooks, although perhaps in opposite directions.

> @sellout please open a separate issue with exact steps on how to reproduce.

Done! #850.
